lib/logic/player/favorites.dart

```diff
@@ -1,27 +1,193 @@
+import 'dart:async';
+
+import 'package:android_content_provider/android_content_provider.dart';
+import 'package:firebase_crashlytics/firebase_crashlytics.dart';
+import 'package:flutter/foundation.dart';
 import 'package:sweyer/sweyer.dart';
 
-// class FavoriteRepository {
-//   FavoriteRepository._();
-//   static final instance = FavoriteRepository._();
+@visibleForTesting
+class FavoritesRepository {
+  final serializersMap = ContentMap<IntSerializerType>({
+    for (final contentType in Content.enumerate())
+      contentType: IntListSerializer('favorites_${ContentUtils.contentTypeId(contentType)}.json')
+  });
+}
+
+class FavoritesControl with Control {
+  static FavoritesControl instance = FavoritesControl();
+
+  @visibleForTesting
+  final repository = FavoritesRepository();
+
+  final _favoriteSetsMap = ContentMap<Set<int>>();
+
+  bool _useMediaStoreFavorites(Type contentType) =>
+    contentType == Song &&
+    DeviceInfoControl.instance.useScopedStorageForFileModifications &&
+    Settings.useMediaStoreForFavoriteSongs.value;
+
+  MediaStoreContentObserver<Song>? _mediaStoreContentObserver;
+
+  @override
+  Future<void> init() async {
+    super.init();
+    _showOnlyFavoritesNotifier.value = false;
+    for (final contentType in Content.enumerate()) {
+      await _initContentType(contentType);
+    }
+    _registerMediaStoreObserver();
+    Settings.useMediaStoreForFavoriteSongs.addListener(_useMediaStoreForFavoriteSongsListener);
+  }
+
+  @override
+  void dispose() {
+    _favoriteSetsMap.clear();
+    _disposeMediaStoreObserver();
+    Settings.useMediaStoreForFavoriteSongs.removeListener(_useMediaStoreForFavoriteSongsListener);
+    super.dispose();
+  }
+
+  Future<void> _initContentType(Type contentType) async {
+    final Set<int> favoriteSet = {};
+    if (_useMediaStoreFavorites(contentType)) {
+      for (final song in ContentControl.instance.state.allSongs.songs) {
+        if (song.isFavoriteInMediaStore!) {
+          favoriteSet.add(song.sourceId);
+        }
+      }
+    } else {
+      final serializer = repository.serializersMap.getValue(contentType);
+      await serializer!.init();
+      final savedIds = await serializer.read();
+      favoriteSet.addAll(savedIds);
+    }
+    _favoriteSetsMap.setValue(favoriteSet, key: contentType);
+  }
+
+  Future<void> _useMediaStoreForFavoriteSongsListener() async {
+    if (Settings.useMediaStoreForFavoriteSongs.value) {
+      _registerMediaStoreObserver();
+    } else {
+      _disposeMediaStoreObserver();
+    }
+    await _initContentType(Song);
+    ContentControl.instance.emitContentChange();
+  }
+
```

Comment:
I don't want to add any logic around this, because you can turn it off, then add a few favorites, then turn it on, and then turn it off again.

Reply:
I don't see why this why you wouldn't want to improve the turning off action, even if the turning on action doesn't do the exact same. It's a first step and just so simple, but the effect is a huge improvement of usability when turning off the integration, you keep all your favorites.

This doesn't intent to fix everything when switching between the settings, it just improves specifically the turning off action, with very little logic.

Reply:
> This doesn't intent to fix everything when switching between the settings, it just improves specifically the turning off action, with very little logic.

But it makes experience worse if you do as I said above, because you lose all the data you could have in the app storage. There's no good heuristic here, unfortunately.